### PR TITLE
Extract cookies

### PR DESCRIPTION
**Pull request by aggroskater**
Howdy folks,

I'm attempting to integrate with a CRM (amoCRM) that expects session tokens to be handled via Cookie headers. The current `header.*` request variable functionality would work for making the Cookie headers on requests to the CRM. However, in this case, the initial setting of the session token is also only available in the Set-Cookie headers of the login response. So far as I know, there is no existing functionality for interrogating response headers (and I can't think of a general way to implement such a feature that would also be useful for further actions in flows, since HTTP headers are largely arbitrary so far as I know).

That being said, Cookie headers in particular are in common use and have specs that try to nail down their behaviour (RFC 6265). I'd like to be able to collect cookies into a named property in the event object if requested, and use their values in subsequent calls. Something like

"header.Cookie": "session_id={{some_previous_destination.some_cookie_holder_property.some_cookie_name.value_raw}}".

This should permit me to integrate with amoCRM.

====================================

I've taken an initial stab at this problem in this pull request. I didn't intend to make a full cookie user agent/cookie jar handler (a bona fide cookie jar would really be more of a flow-level concept). Rather, I just wanted to conveniently expose Set-Cookie header data from a response into an object for later use in the flow. Still, I've included handling for various behaviours that run against RFC 6265 and other edge cases (malformed encodings).

The spec and code assume that `res.headers['Set-Cookie']` is an array of cookie strings. Strictly speaking I don't have visibility into the caller of response() (that is presumably building `res`), but the "Technical Details" of destination requests/responses for any given run through a flow show the Set-Cookie header as such an array. I utilize the set-cookie-parser module to parse each cookie (it's a small one-file module with no dependencies). I then process each cookie to handle RFC-breaking behaviour. And finally I format the cookie into the desired object format in the event.

Let me know what y'all think.

### CODE REVIEW

**Comment by cgrayson** [comment]
Hey, Preston. First, I apologize, we let this fall through the cracks until rediscovering it today. This code is great; we really appreciate the effort you put into the details of the specs. Alex and I discussed the general use-case, and we think it might be able to be simplified.

One thing we wondered is how necessary it is to break the cookie data down to component pieces. Skimming the AmoCRM docs, it looks like they want the whole cookie string header passed back, which is consistent with other services we've seen in the past. Most if not all of those cases are like yours: a session cookie that's issued and has to be returned with subsequent requests, just like a browser client. So, if we treat the cookie value as the whole string, there's no need to break it down or reconstitute it later (or worry about whether it was malformed per RFCs & specs, for that matter). 

Another issue is how to handle multiple cookies. As I'm sure you know, individual entries in an appended array can be referenced in LeadConduit mappings, but only by position (e.g., `cookie.0`). Since this array's order may be non-deterministic, we're worried about the brittleness of relying on that kind of addressing. Instead, we're thinking it would be simpler and more straightforward to ensure there's only ever a single cookie appended by this integration. The problem – illustrated very clearly thanks to your test coverage here! – is that cookies are weird. :-) We're thinking that if we let the user specify a regex matcher, that supports both a simple just-the-name case, as well as a more complex name-plus-domain-or-path-or-whatever scenario.

In summary, here's the kind of interface we came down to:

1. This `custom` integration would get a new request variable, something like `cookie_search_term`. When present, the response-parsing would extract the last cookie, as a string, from the `Set-Cookie` array that string-matches the mapped term.

2. It would also get a new response variable, probably called `cookie` :-) that would be set by this as an appended value.

3. Subsequent recipient steps would be able to map that appended value to `header.Cookie`, just as you described. 

Let me know what you think.

**Comment by aggroskater** [comment]
I have no objections to that approach. It certainly simplifies things (and lets you drop the new set-cookie-parser dependency too). The only hiccup I could see is if the response *also* had a "cookie" property, and then we clobbered it with the value from the headers. Taking the `capture.*` approach, with a named property to go along with the regex, would address that though (if it turned out to actually be an issue), and would also permit capturing multiple cookie strings. On the other hand, leaving it as just a non-wildcard `cookie_search_term` variable to capture just a single cookie is simpler and definitely the most common use-case.

And of course if there were ever a need for more fine-grained parsing and interrogation, well, there's always the original PR code to look to for inspiration.

**Comment by cgrayson** [comment]
Hey, Preston. It took a while for us to get to this, but it's in production now (v 2.14.1). Custom recipients have a new mapping: "Cookie Search Term". As discussed, setting that will capture the cookie that matches it. In the case of multiple matches, it takes the first match from the list sorted lexicographically.

**Comment by aggroskater** [comment]
Thanks so much Chris. I've tested the feature, and it works as expected. I'm now able to integrate with amoCRM :D Y'all're the best!

**Comment by aggroskater** [comment]
Thanks so much Chris. I've tested the feature, and it works as expected. I'm now able to integrate with amoCRM :D Y'all're the best!
